File: logger/logger.go
package logger

import (
	"encoding/json"
	"io"
	"os"
	"sync"
	"time"
)

// Logger writes structured JSON log entries to an io.Writer.
// Each call to Log produces one JSON object per line on the writer.
type Logger struct {
	writer io.Writer
	mu     sync.Mutex // protects concurrent writes
}

// New creates a Logger that writes to the given writer.
func New(w io.Writer) *Logger {
	return &Logger{writer: w}
}

// NewStdout creates a Logger that writes to os.Stdout.
func NewStdout() *Logger {
	return New(os.Stdout)
}

// Log writes a single JSON log entry. It automatically adds a "timestamp"
// field with the current time in RFC3339 format. The fields map is merged
// into the output — any key called "timestamp" in fields will be overwritten.
func (l *Logger) Log(fields map[string]any) {
	l.mu.Lock()
	defer l.mu.Unlock()

	// Add timestamp
	fields["timestamp"] = time.Now().UTC().Format(time.RFC3339)

	data, err := json.Marshal(fields)
	if err != nil {
		// Fallback: log the error itself
		fallback, _ := json.Marshal(map[string]any{
			"timestamp": time.Now().UTC().Format(time.RFC3339),
			"event":     "log_error",
			"error":     err.Error(),
		})
		l.writer.Write(fallback)
		l.writer.Write([]byte("\n"))
		return
	}

	l.writer.Write(data)
	l.writer.Write([]byte("\n"))
}

File: src/logger/logger.ts
import type { Writable } from "node:stream";

/**
 * Logger writes structured JSON log entries to a writable stream.
 * Each call to log() produces one JSON object per line.
 */
export class Logger {
  private writer: Writable;

  constructor(writer: Writable) {
    this.writer = writer;
  }

  /**
   * Writes a single JSON log entry. Automatically adds a "timestamp"
   * field with the current time in ISO 8601 (RFC 3339) format.
   */
  log(fields: Record<string, unknown> | null): void {
    const entry: Record<string, unknown> = {
      ...(fields ?? {}),
      timestamp: new Date().toISOString(),
    };

    try {
      const data = JSON.stringify(entry);
      this.writer.write(data + "\n");
    } catch {
      // Fallback: log the serialization error itself
      const fallback = JSON.stringify({
        timestamp: new Date().toISOString(),
        event: "log_error",
        error: "failed to serialize log entry",
      });
      this.writer.write(fallback + "\n");
    }
  }
}

/** Creates a Logger that writes to process.stdout. */
export function newStdoutLogger(): Logger {
  return new Logger(process.stdout);
}
